In [100]:
import pandas as pd


In [45]:
retail_df = pd.read_excel("C:\\Users\\sheryl.ojwando\\Downloads\\Online Retail.xlsx")

In [41]:
retail_df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


### Data cleaning and preprocessing

In [42]:
retail_df.shape

(541909, 8)

In [43]:
retail_df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [34]:
retail_df.nunique()

InvoiceNo      22190
StockCode       3684
Description     3896
Quantity         436
InvoiceDate    20460
UnitPrice        620
CustomerID      4372
Country           37
dtype: int64

In [35]:
retail_df.duplicated().sum()

np.int64(5225)

In [37]:
## Drop columns where there is no customer id
## in pandas dropna drops columns where the value is null( Nan)
retail_df1 = retail_df.dropna(subset = ['CustomerID'])
retail_df1.isnull().sum()




                                                        
                                                        
                                    
                                                        

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [38]:
retail_df1.shape

(406829, 8)

In [47]:
retail_df1.nunique()

InvoiceNo      22190
StockCode       3684
Description     3896
Quantity         436
InvoiceDate    20460
UnitPrice        620
CustomerID      4372
Country           37
dtype: int64

In [55]:
#Check the duplicated sum of values
retail_df1.duplicated().sum()

np.int64(5225)

In [59]:
## show the values that are duplicated
retail_df1.duplicated()
retail_df1[retail_df1.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
...,...,...,...,...,...,...,...,...
541675,581538,22068,BLACK PIRATE TREASURE CHEST,1,2011-12-09 11:34:00,0.39,14446.0,United Kingdom
541689,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,2011-12-09 11:34:00,2.49,14446.0,United Kingdom
541692,581538,22992,REVOLVER WOODEN RULER,1,2011-12-09 11:34:00,1.95,14446.0,United Kingdom
541699,581538,22694,WICKER STAR,1,2011-12-09 11:34:00,2.10,14446.0,United Kingdom


In [61]:
## Keep the first occurence of duplicates, delete the rest
retail_df1= retail_df1.drop_duplicates()

In [64]:
retail_df1.duplicated().sum()

np.int64(0)

In [66]:
retail_df1.shape


(401604, 8)

In [67]:
retail_df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  float64       
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.6+ MB


In [69]:
## Transform the CustomerID data type
retail_df1['CustomerID'] = retail_df1['CustomerID'].astype(int)

In [70]:
retail_df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  int64         
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 27.6+ MB


In [73]:
## Do a summary statistics
retail_df1[['InvoiceDate', 'Quantity', 'UnitPrice']].describe()

,InvoiceDate,Quantity,UnitPrice
count,401604,401604.000000,401604.000000
mean,2011-07-10 12:08:23.848567552,12.183273,3.474064
min,2010-12-01 08:26:00,-80995.000000,0.000000
25%,2011-04-06 15:02:00,2.000000,1.250000
50%,2011-07-29 15:40:00,5.000000,1.950000
75%,2011-10-20 11:58:30,12.000000,3.750000
max,2011-12-09 12:50:00,80995.000000,38970.000000
std,NaN,250.283037,69.764035


#### As we can see from this summary statistics, the invoice date runs from 1st dec 2010 to 1st dec 2011.
#### There are negative values for quantity, which we must investigate
#### There are unit prices with 0 values

In [74]:
retail_df1[['Description','Country']].describe()

,Description,Country
count,401604,401604
unique,3896,37
top,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,2058,356728


#### The UK is the leading country, and the top selling product is WHITE HANGING HEART T-LIGHT HOLDER

In [79]:
## check if the Country names have been standardized
countries= retail_df1['Country'].unique()
countries

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Greece', 'Singapore', 'Lebanon',
       'United Arab Emirates', 'Saudi Arabia', 'Czech Republic', 'Canada',
       'Unspecified', 'Brazil', 'USA', 'European Community', 'Bahrain',
       'Malta', 'RSA'], dtype=object)

In [89]:
## Create a filter to remove rows with -tive value for quantity and 0 for unit price
retail_df_clean = retail_df1[(retail_df1['Quantity']>= 1) & (retail_df1['UnitPrice']> 0)].copy()

In [91]:
retail_df_clean.shape

(392692, 8)

In [92]:
## check if the description names have been standardized
desc= retail_df1['Description'].unique()
desc

array(['WHITE HANGING HEART T-LIGHT HOLDER', 'WHITE METAL LANTERN',
       'CREAM CUPID HEARTS COAT HANGER', ...,
       'PINK CRYSTAL SKULL PHONE CHARM',
       'CREAM HANGING HEART T-LIGHT HOLDER',
       'PAPER CRAFT , LITTLE BIRDIE'], dtype=object)

In [93]:
### Run a summary statistics of the cleaned dataset
retail_df_clean[['InvoiceDate', 'Quantity', 'UnitPrice']].describe()

,InvoiceDate,Quantity,UnitPrice
count,392692,392692.000000,392692.000000
mean,2011-07-10 19:13:07.771892480,13.119702,3.125914
min,2010-12-01 08:26:00,1.000000,0.001000
25%,2011-04-07 11:12:00,2.000000,1.250000
50%,2011-07-31 12:02:00,6.000000,1.950000
75%,2011-10-20 12:53:00,12.000000,3.750000
max,2011-12-09 12:50:00,80995.000000,8142.750000
std,NaN,180.492832,22.241836


In [95]:
## Create the revenue column
retail_df_clean['Revenue']= retail_df_clean['Quantity'] * retail_df_clean['UnitPrice']
retail_df_clean.head()


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


In [96]:
### Run a summary statistics of the cleaned dataset
retail_df_clean[['InvoiceDate', 'Quantity', 'UnitPrice','Revenue']].describe()

,InvoiceDate,Quantity,UnitPrice,Revenue
count,392692,392692.000000,392692.000000,392692.000000
mean,2011-07-10 19:13:07.771892480,13.119702,3.125914,22.631500
min,2010-12-01 08:26:00,1.000000,0.001000,0.001000
25%,2011-04-07 11:12:00,2.000000,1.250000,4.950000
50%,2011-07-31 12:02:00,6.000000,1.950000,12.450000
75%,2011-10-20 12:53:00,12.000000,3.750000,19.800000
max,2011-12-09 12:50:00,80995.000000,8142.750000,168469.600000
std,NaN,180.492832,22.241836,311.099224


#### Rows: 392,692 (Cleaned and verified).

#### Columns: All original columns plus the new Revenue column.

#### Quality: No negatives, no zeroes, and no missing customer identifiers.

In [99]:
# Save the file directly to the Downloads folder
file_path = r"C:\Users\sheryl.ojwando\Downloads\Online_Retail_PowerBI.csv"

retail_df_clean.to_csv(file_path, index=False, encoding='utf-8-sig')

print(f"Success! The file has been saved to: {file_path}")

Success! The file has been saved to: C:\Users\sheryl.ojwando\Downloads\Online_Retail_PowerBI.csv
